### PR TITLE
feat: Use sentry-go

### PR DESCRIPTION
**Pull request by AbhiPrasad**
## Motivation

Now that https://github.com/getsentry/sentry-go/pull/235 has been merged in, we should implement the transactions from `sentry-go` with the collector. 

This PR aims to generate transactions from `rootSpanTrees`, and then attempt to send them to sentry using `sentry-go`'s async transport.

## Implementation.

Please view https://github.com/getsentry/opentelemetry-collector-contrib/blob/6cd4cc16f2807b9d187e950e9cb381537cbb1e52/exporter/sentryexporter/docs/transformation.md on more details on how we chose to generate `rootSpanTrees`.

A root span tree is a datastructure that represents a Sentry transactions. It is made up of the following fields:

```
type rootSpanTree struct {
	rootSpan       *sentry.Span
	childSpans     []*sentry.Span
	libraryName    string
	libraryVersion string
	resourceTags   map[string]string
}
```

From a `rootSpanTree`, a transaction can easily be generated. We then can take that transaction, and use the `sentry-go` `HTTPTransport` to send it to Sentry.

### CODE REVIEW

**Comment by rhcarvalho** [comment]
omg, it starts off with already 57 commits! Big baby.

**Comment by AbhiPrasad** [comment]
Yes, this is due to the fact that it's built on top of #5

File: go.sum [hunk 1]
```diff
@@ -28,6 +28,7 @@ contrib.go.opencensus.io/exporter/stackdriver v0.13.1/go.mod h1:z2tyTZtPmQ2HvWH4
 contrib.go.opencensus.io/resource v0.1.2 h1:b4WFJV8u7/NzPWHeTqj3Ec2AW8OGhtJxC/hbphIOvbU=
 contrib.go.opencensus.io/resource v0.1.2/go.mod h1:F361eGI91LCmW1I/Saf+rX0+OFcigGlFvXwEGEnkRLA=
 dmitri.shuralyov.com/gpu/mtl v0.0.0-20190408044501-666a987793e9/go.mod h1:H6x//7gZCb22OMCxBHrMx7a5I7Hp++hsVxbQ4BYO7hU=
+github.com/AndreasBriese/bbloom v0.0.0-20190306092124-e2d15f34fcf9/go.mod h1:bOvUY6CB00SOBii9/FifXqc0awNKxLFCL/+pkDPuyl8=
```
Comment (rhcarvalho) [inline go.sum:31]: @AbhiPrasad please run `go mod tidy` on the root of the repo, these changes to `/go.mod` were probably accidental. Adding the new exporter should not change the top level module, right?

File: exporter/sentryexporter/sentry_exporter.go [hunk 2]
```diff
@@ -99,11 +129,19 @@ func (s *SentryExporter) pushTraceData(ctx context.Context, td pdata.Traces) (dr
...
+				// keep track of that transaction.
+				//
+				// If the span is not a root span, we can either associate it with an existing
+				// span tree, or we can temporarily consider it an orphan span.
+				if IsRootSpan(sentrySpan) {
 					rootSpanTreeMap[sentrySpan.SpanID] = &rootSpanTree{
-						rootSpan:   sentrySpan,
-						childSpans: make([]*SentrySpan, 0),
+						rootSpan:       sentrySpan,
+						childSpans:     make([]*sentry.Span, 0),
```
Comment (rhcarvalho) [inline exporter/sentryexporter/sentry_exporter.go:141]: You can remove all lines that contain `make([]*sentry.Span, 0)` in the code base, for what I could observe all cases are redundant allocations.

Note that to change `childSpans` / add a `Span` we write

```go
root.childSpans = append(root.childSpans, someSpan)
```

Go's `append` is designed such that most of the time you don't need to initialize memory for a slice, `append` does it for you (this is part of the principle of making the zero-value useful -- a `nil` slice is ready for appending):

```go
var s []string
s = append(s, "go")    // this is equivalent to s := append(nil, "go")
```

---

The common reason to use `make` to initialize a slice is when you know it's final length. E.g. if you have a `[]int` and you need a `[]float64`:

```go
func convert(s []int) []float64 {
	r := make([]float64, len(s))
	for i := 0; i < len(s); i++ {
		r[i] = float64(s[i])
	}
	return r
}
```

And if it is important to avoid unnecessary allocations in the example `convert` function above, then the code could also check for `len(s) == 0` or `s == nil` and return earlier.

File: exporter/sentryexporter/sentry_exporter.go [hunk 3]
```diff
@@ -49,28 +52,40 @@ var (
...
+
+type spanCollection struct {
+	span           *sentry.Span
+	libraryName    string
+	libraryVersion string
+	resourceTags   map[string]string
 }
 
-// TODO: Add to function
 func (s *SentryExporter) pushTraceData(ctx context.Context, td pdata.Traces) (droppedSpans int, err error) {
```
Comment (rhcarvalho) [inline exporter/sentryexporter/sentry_exporter.go:81]: Possibly out of scope for this PR, but since it affects reading/reviewing the code, I'll mention anyway.

This function has too many levels of indentation, making it hard to follow.
We could look into how to break up the 3 nested loops into smaller functions for readability.

File: exporter/sentryexporter/sentry_exporter.go [hunk 4]
```diff
@@ -83,13 +98,28 @@ func (s *SentryExporter) pushTraceData(ctx context.Context, td pdata.Traces) (dr
...
+			libName := ""
+			libVersion := ""
+			if !library.IsNil() {
+				name := library.Name()
+				version := library.Version()
+
+				if name != "" && version != "" {
+					libName = name
+					libVersion = version
+				}
```
Comment (rhcarvalho) [inline exporter/sentryexporter/sentry_exporter.go:120]: Why do we require both `name` and `version` to be simultaneously non-empty?

Wouldn't it work to assign the values from `library` directly?

```go
libName = library.Name()
libVersion = library.Version()
```

File: exporter/sentryexporter/sentry_exporter.go [hunk 5]
```diff
@@ -49,28 +52,40 @@ var (
...
 	}
+
+	otelSentryExporterVersion = "0.0.1"
+	otelSentryExporterName    = "sentry.opentelemetry.collector"
 )
 
 // SentryExporter defines the Sentry Exporter.
 type SentryExporter struct {
-	DSN string
+	transport *sentry.HTTPTransport
```
Comment (rhcarvalho) [inline exporter/sentryexporter/sentry_exporter.go:62]: Did you consider using the `HTTPSyncTransport`?

The tight loop calling `s.transport.SendEvent()` will drop transactions if there is more than 30 transactions in a batch. If not using the Sync transport, I'd limit sending a few transactions, blocking on a call to Flush, then continuing to the next batch (doesn't really solve running into upstream rate limiting, but at least doesn't drop events right at the SDK transport level).

File: exporter/sentryexporter/sentry_exporter.go [hunk 6]
```diff
@@ -49,28 +52,40 @@ var (
...
-	rootSpan   *SentrySpan
-	childSpans []*SentrySpan
+	rootSpan       *sentry.Span
+	childSpans     []*sentry.Span
+	libraryName    string
+	libraryVersion string
+	resourceTags   map[string]string
+}
+
+type spanCollection struct {
```
Comment (rhcarvalho) [inline exporter/sentryexporter/sentry_exporter.go:74]: This name here is misleading. When I saw it in use in func signatures I was expecting "more than one span", but this seems to really the a single span with added metadata.

Instead of proposing a new name, I wonder if this type is necessary at all. Why can't we store `libraryName`, `libraryVersion` and `resourceTags` directly in the `sentry.Span` value?

File: exporter/sentryexporter/sentry_exporter.go [hunk 7]
```diff
@@ -49,28 +52,40 @@ var (
 		"data_loss",
 		"unauthenticated",
 	}
+
+	otelSentryExporterVersion = "0.0.1"
+	otelSentryExporterName    = "sentry.opentelemetry.collector"
```
[the remark this answers is not in the document]
Comment (AbhiPrasad) [reply exporter/sentryexporter/sentry_exporter.go:57]: Changed to `sentry.opentelemetry`, and regrouped the variables.

File: exporter/sentryexporter/sentry_exporter.go (hunk 4 above)
Comment (AbhiPrasad) [reply exporter/sentryexporter/sentry_exporter.go:120]: Yes that was left over from previous testing, will change.

File: exporter/sentryexporter/sentry_exporter.go (hunk 5 above)
Comment (AbhiPrasad) [reply exporter/sentryexporter/sentry_exporter.go:62]: Yes initially I used HTTPSyncTransport, but I found it was a bottleneck when I did synthetic load testing with https://github.com/Omnition/synthetic-load-generator, so I switched over to using the async transport.

Great point about the 30 transactions min though, I will address this.

File: exporter/sentryexporter/sentry_exporter.go (hunk 6 above)
Comment (AbhiPrasad) [reply exporter/sentryexporter/sentry_exporter.go:74]: So I did this because I wanted to avoid redundant data, but it might be more correct to not do this because a transaction can be made up of spans from different libraries and different resources. Will remove this concept and just add it as `sentry.Span` data.

File: exporter/sentryexporter/interfaces.go [hunk 8]
```diff
@@ -14,48 +14,45 @@
...
+		SpanID:      rtree.rootSpan.SpanID,
+		Op:          rtree.rootSpan.Op,
+		Description: rtree.rootSpan.Description,
+		Status:      rtree.rootSpan.Status,
+	}
+
+	transaction.Contexts["library"] = map[string]string{
+		"name":    rtree.libraryName,
+		"version": rtree.libraryVersion,
+	}
```
[the remark this answers is not in the document]
Comment (AbhiPrasad) [reply exporter/sentryexporter/interfaces.go:39]: Switching to using field data.

File: exporter/sentryexporter/sentry_exporter.go (hunk 3 above)
Comment (AbhiPrasad) [reply exporter/sentryexporter/sentry_exporter.go:81]: Yes agree, will address in the future.

File: exporter/sentryexporter/sentry_exporter.go (hunk 5 above)
Comment (AbhiPrasad) [reply exporter/sentryexporter/sentry_exporter.go:62]: Addressed, also added transport buffer size as a config option.

**Comment by AbhiPrasad** [comment]
Update: Addressed PR Review comments:

* Now set library name/version and resource tags on span, instead of carrying it around in a `spanCollection`.
* Made `IsRootSpan` -> `isRootSpan`
* Added transport buffer size config option
* Make sure to flush transactions before they overflow the transport buffer

File: exporter/sentryexporter/sentry_exporter.go (hunk 5 above)
Comment (rhcarvalho) [reply exporter/sentryexporter/sentry_exporter.go:62]: > I found it was a bottleneck when I did synthetic load testing with https://github.com/Omnition/synthetic-load-generator, so I switched over to using the async transport

Maybe the "solution" became really "drop a lot of events". When you set out to test this, make sure you check the events you send on one end show up on the other end :)

File: exporter/sentryexporter/sentry_exporter.go [hunk 9]
```diff
@@ -112,45 +126,84 @@ func (s *SentryExporter) pushTraceData(ctx context.Context, td pdata.Traces) (dr
...
+		// buffer size so we don't drop transactions.
+		if bufferCounter == transport.BufferSize {
+			transport.Flush(time.Second)
+			bufferCounter = 0
+		}
+
+		transport.SendEvent(t)
+		bufferCounter++
+	}
+}
```
Comment (rhcarvalho) [inline exporter/sentryexporter/sentry_exporter.go:161]: Note: this works around the transport's buffer, but doesn't deal with rate limits.
Maybe we've talked already about shaping the traffic with a [leaky bucket](https://en.wikipedia.org/wiki/Leaky_bucket) / [token bucket](https://en.wikipedia.org/wiki/Token_bucket), but that would only work if you have a single point of ingestion for a given DSN.

As Bruno mentioned, we have to understand the bigger picture, how is the data flowing through the exporter, how are spans batched, ...

The current implementation might be good for starters, though still more complex than if we used `sentry.HTTPSyncTransport` (which I think you ruled out for bad performance?).

File: exporter/sentryexporter/config.go [hunk 10]
```diff
@@ -21,4 +21,6 @@ type Config struct {
 	configmodels.ExporterSettings `mapstructure:",squash"` // squash ensures fields are correctly decoded in embedded struct.
 	// DSN to report transaction to Sentry. If the DSN is not set, no trace will be sent to Sentry.
 	DSN string `mapstructure:"dsn"`
+	// Buffer size for Sentry transport. Defaults to 30.
```
Comment (rhcarvalho) [inline exporter/sentryexporter/config.go:24]: The tricky thing about exposing this is that we don't document what buffer sizes to use and the consequences.

We may want not just a buffer, but also multiple concurrent requests (which puts you closer to hitting a rate limit), but could be controlled.

For now the `HTTPTransport` has a buffer that only affects memory consumption, still all requests are sent sequentially.

How about we keep the configuration internal for now, and add this once we have more feedback?

File: exporter/sentryexporter/sentry_exporter.go [hunk 11]
```diff
@@ -83,27 +88,36 @@ func (s *SentryExporter) pushTraceData(ctx context.Context, td pdata.Traces) (dr
...
+				sentrySpan := convertToSentrySpan(otelSpan, library, resourceTags)
 
-				if sentrySpan.IsRootSpan() {
-					// Add root span to span store map
+				// If a span is a root span, we consider it the start of a Sentry transaction.
+				// We should then keep create a new root span tree for that root span, and
+				// keep track of that transaction.
+				//
+				// If the span is not a root span, we can either associate it with an existing
+				// span tree, or we can temporarily consider it an orphan span.
```
[the remark this answers is not in the document]
Comment (AbhiPrasad) [reply exporter/sentryexporter/sentry_exporter.go]: Good idea. I refactored to replace root span tree with just transactions.

File: exporter/sentryexporter/sentry_exporter.go (hunk 3 above)
Comment (AbhiPrasad) [reply exporter/sentryexporter/sentry_exporter.go:66]: Hmm. Now after multiple rounds of refactoring, the `pushTraceData` is much smaller as a function. I feel that it is fine for the function to be at the size it is, and be explicit about what exactly it is doing.

File: exporter/sentryexporter/factory.go [hunk 12]
```diff
@@ -43,6 +43,7 @@ func (f *Factory) CreateDefaultConfig() configmodels.Exporter {
 			TypeVal: typeStr,
 			NameVal: typeStr,
 		},
+		DSN: "",
```
Comment (rhcarvalho) [inline exporter/sentryexporter/factory.go:46]: Empty string is the zero-value for strings, `DSN: ""` is the same as omitting this line (unless you want to explicitly list out all config options).
